File: 2_Creational_Design_Patterns/2.03_Prototype/Java_Snippet/Prototype.java
import java.util.ArrayList;
import java.util.List;

// Prototype interface for cakes
interface CakePrototype {
    CakePrototype clone();

    void describe();
}

// Concrete prototype for a basic cake
class BasicCake implements CakePrototype {
    private String flavor;
    private int layers;

    public BasicCake(String flavor, int layers) {
        this.flavor = flavor;
        this.layers = layers;
    }

    @Override
    public CakePrototype clone() {
        return new BasicCake(flavor, layers);
    }

    @Override
    public void describe() {
        System.out.println("A basic " + flavor + " cake with " + layers + " layers");
    }
}

// Concrete prototype for a decorated cake
class DecoratedCake implements CakePrototype {
    private CakePrototype baseCake;
    private List<String> decorations;

    public DecoratedCake(CakePrototype baseCake, List<String> decorations) {
        this.baseCake = baseCake;
        this.decorations = new ArrayList<>(decorations);
    }

    @Override
    public CakePrototype clone() {
        List<String> clonedDecorations = new ArrayList<>(decorations);
        return new DecoratedCake(baseCake.clone(), clonedDecorations);
    }

    @Override
    public void describe() {
        System.out.println("A decorated cake:");
        baseCake.describe();
        System.out.println("Decorations: " + decorations);
    }
}

// Example usage
public class Prototype {
    public static void main(String[] args) {
        BasicCake basicChocolateCake = new BasicCake("chocolate", 2);
        BasicCake basicVanillaCake = new BasicCake("vanilla", 3);

        // Cloning a basic cake and adding decorations
        List<String> decorations = new ArrayList<>();
        decorations.add("chocolate shavings");
        decorations.add("strawberries");
        DecoratedCake decoratedChocolateCake = new DecoratedCake(basicChocolateCake, decorations);

        basicChocolateCake.describe();
        decoratedChocolateCake.describe();
    }
}

File: 2_Creational_Design_Patterns/2.03_Prototype/Python_Snippet/Prototype.py
import copy

# Prototype class for cakes
class CakePrototype:
    def clone(self):
        # This method will be overridden by subclasses
        pass

    def describe(self):
        # This method will be overridden by subclasses
        pass

# Concrete prototype for a basic cake
class BasicCake(CakePrototype):
    def __init__(self, flavor, layers):
        self.flavor = flavor
        self.layers = layers

    def clone(self):
        return copy.deepcopy(self)

    def describe(self):
        print(f"A basic {self.flavor} cake with {self.layers} layers")

# Concrete prototype for a decorated cake
class DecoratedCake(CakePrototype):
    def __init__(self, base_cake, decorations):
        self.base_cake = base_cake
        self.decorations = decorations

    def clone(self):
        # Cloning the base cake and then adding decorations
        cloned_cake = self.base_cake.clone()
        cloned_cake.decorations = self.decorations
        return cloned_cake

    def describe(self):
        print("A decorated cake:")
        self.base_cake.describe()
        print("Decorations:", self.decorations)

# Example usage
basic_chocolate_cake = BasicCake("chocolate", 2)
basic_vanilla_cake = BasicCake("vanilla", 3)

# Cloning a basic cake and adding decorations
decorated_chocolate_cake = DecoratedCake(basic_chocolate_cake, ["chocolate shavings", "strawberries"])

basic_chocolate_cake.describe()
decorated_chocolate_cake.describe()